<h2><font color="#004D7F" size=6>Módulo 2. Análisis de datos</font></h2>



<h1><font color="#004D7F" size=5>2. Estadística descriptiva</font></h1>

<br><br>
<div style="text-align: right">
<font color="#004D7F" size=3>Manuel Castillo-Cara</font><br>
<font color="#004D7F" size=3>Machine Learning con Python</font><br>

---

<h2><font color="#004D7F" size=5>Índice</font></h2>
<a id="indice"></a>

* [1. Introducción](#section1)
    * [1.1. Cargar el conjunto de datos](#section11)
* [2. Funciones de estadística descriptiva](#section2)
    * [2.1. Revisar los datos: _head()_](#section21)
    * [2.2. Dimensiones de los datos: _shape_](#section22)
    * [2.3. Tipo de datos: _dtypes_](#section23)
    * [2.4. Resumen: _describe()_](#section24)
    * [2.5. Distribución entre clases: _groupby('class').size()_](#section25)
    * [2.6. Correlaciones: _corr()_](#section26)
    * [2.7. Asimetría: _skew()_](#section27)

In [3]:
# Permite ajustar la anchura de la parte útil de la libreta (reduce los márgenes)
from IPython.core.display import display, HTML
display(HTML("<style>.container{ width:98% }</style>"))

---

<a id="section1"></a>
# <font color="#004D7F"> 1. Introducción</font>

Unos de los errores principales que se cometen cuando se comienza a trabajar con machine learning es tomar decisiones directamente a través de los algoritmos sin un análisis previo del conjunto de datos a trabajar. Es importante que entendamos que, un análisis de datos y un buen preprocesamiento de los mismos antes de realizar un tratamiento de modelado, nos llevará no solamente a obtener mejores y más confiables resultados, sino que entenderemos a la perfección nuestro conjunto de datos dando una gran experiencia en el tiempo en la fase de análisis y tratamiento de datos.

<a id="section11"></a>
## <font color="#004D7F">1.1. Cargar el conjunto de datos</font>

Para esta práctica vamos a cargar el conjunto de datos de Pima Indian Diabetes para hacer observaciones con las funciones de estadística descriptiva.

In [4]:
import pandas as pd 
filename = 'data/pima-indians-diabetes.csv'
names = ['preg','plas','pres','skin','test','mass','pedi','age','class']
data = pd.read_csv(filename, names=names)
data

preg  plas  pres  skin  test  mass     pedi  age  class
0       6   148    72    35     0  33.6   627.00   50      1
1       1    85    66    29     0  26.6   351.00   31      0
2       8   183    64     0     0  23.3   672.00   32      1
3       1    89    66    23    94  28.1   167.00   21      0
4       0   137    40    35   168  43.1  2288.00   33      1
..    ...   ...   ...   ...   ...   ...      ...  ...    ...
763    10   101    76    48   180  32.9   171.00   63      0
764     2   122    70    27     0  36.8     0.34   27      0
765     5   121    72    23   112  26.2   245.00   30      0
766     1   126    60     0     0  30.1   349.00   47      1
767     1    93    70    31     0  30.4   315.00   23      0

[768 rows x 9 columns]

<div style="text-align: right"> <font size=5>
    <a href="#indice"><i class="fa fa-arrow-circle-up" aria-hidden="true" style="color:#004D7F"></i></a>
</font></div>

---

<a id="section2"></a>
# <font color="#004D7F"> 2. Funciones de estadística descriptiva</font>

La inspección de nuestro conjunto de datos es fundamental para poder entender mejor que técnica utilizar; además, nos ayudará a desarrollar nuestra intuición y hacernos preguntas sobre ellos. Las múltiples perspectivas de sus datos lo desafiarán a pensar en los datos de manera diferente.

<a id="section21"></a>
## <font color="#004D7F">2.1. Revisar los datos: _head()_ </font>

Puede revisar las primeras 20 filas de sus datos utilizando la función `head()` en el DataFrame de Pandas. Puede ver que la primera columna enumera el número de fila, lo cual es útil para hacer referencia a una observación específica.

In [5]:
data.head(20)

preg  plas  pres  skin  test  mass    pedi  age  class
0      6   148    72    35     0  33.6   627.0   50      1
1      1    85    66    29     0  26.6   351.0   31      0
2      8   183    64     0     0  23.3   672.0   32      1
3      1    89    66    23    94  28.1   167.0   21      0
4      0   137    40    35   168  43.1  2288.0   33      1
5      5   116    74     0     0  25.6   201.0   30      0
6      3    78    50    32    88  31.0   248.0   26      1
7     10   115     0     0     0  35.3   134.0   29      0
8      2   197    70    45   543  30.5   158.0   53      1
9      8   125    96     0     0   0.0   232.0   54      1
10     4   110    92     0     0  37.6   191.0   30      0
11    10   168    74     0     0  38.0   537.0   34      1
12    10   139    80     0     0  27.1  1441.0   57      0
13     1   189    60    23   846  30.1   398.0   59      1
14     5   166    72    19   175  25.8   587.0   51      1
15     7   100     0     0     0  30.0   484.0   32      1
16     0   118    84    47   230  45.8   551.0   31      1
17     7   107    74     0     0  29.6   254.0   31      1
18     1   103    30    38    83  43.3   183.0   33      0
19     1   115    70    30    96  34.6   529.0   32      1

<a id="section22"></a>
## <font color="#004D7F">2.2. Dimensiones de los datos: _shape_ </font>

Puede revisar la forma y el tamaño de su conjunto de datos imprimiendo la propiedad `shape` en el DataFrame de Pandas. Los resultados se enumeran en filas y luego en columnas. Puede ver que el conjunto de datos tiene 768 filas y 9 columnas

In [6]:
# Dimensions of your data
data.shape


(768, 9)

<a id="section23"></a>
## <font color="#004D7F">2.3. Tipo de datos: _dtypes_ </font>

Puede enumerar los tipos de datos utilizados por el DataFrame para caracterizar cada atributo utilizando la propiedad `dtypes`. Puede ver que la mayoría de los atributos son enteros y que `mass` y `pedi` son tipos de coma flotante.

In [7]:
# Data Types for Each Attribute
data.dtypes

preg       int64
plas       int64
pres       int64
skin       int64
test       int64
mass     float64
pedi     float64
age        int64
class      int64
dtype: object

<a id="section24"></a>
## <font color="#004D7F">2.4. Resumen: _describe()_ </font>

Puedes ver que obtienes muchos datos. Notará algunas llamadas como `pandas.set_option()` utilizada para cambiar la precisión de los números y el ancho preferido de la salida. Esto es para que sea más legible para este ejemplo. Al describir sus datos de esta manera, vale la pena tomarse un tiempo y revisar las observaciones de los resultados.

In [13]:
# Statistical Summary
pd.set_option('display.width', 100)
#pd.set_option('precision', 3)
data.describe()


preg        plas        pres        skin        test        mass         pedi  \
count  768.000000  768.000000  768.000000  768.000000  768.000000  768.000000   768.000000   
mean     3.845052  120.894531   69.105469   20.536458   79.799479   31.992578   428.235091   
std      3.369578   31.972618   19.355807   15.952218  115.244002    7.884160   340.485655   
min      0.000000    0.000000    0.000000    0.000000    0.000000    0.000000     0.100000   
25%      1.000000   99.000000   62.000000    0.000000    0.000000   27.300000   205.000000   
50%      3.000000  117.000000   72.000000   23.000000   30.500000   32.000000   337.000000   
75%      6.000000  140.250000   80.000000   32.000000  127.250000   36.600000   591.500000   
max     17.000000  199.000000  122.000000   99.000000  846.000000   67.100000  2329.000000   

              age       class  
count  768.000000  768.000000  
mean    33.240885    0.348958  
std     11.760232    0.476951  
min     21.000000    0.000000  
25%     24.000000    0.000000  
50%     29.000000    0.000000  
75%     41.000000    1.000000  
max     81.000000    1.000000

<a id="section25"></a>
## <font color="#004D7F">2.5. Distribución entre clases: _groupby('class').size()_ </font>

Puede ver que hay casi el doble de observaciones con la clase 0 (sin aparición de diabetes) que con la clase 1 (aparición de diabetes). En este caso podemos observar que las clases se encuentran desbalanceadas por lo que los tenemos que analizar muy los resultados de los algoritmos.

In [15]:
# Class Distribution
# problema de clasificacion (binaria o multiclase)  ### NO FUNCIONA CON REGRESION###
data.groupby('class').size()

# 100
# 95 -> 0
# 5 -> 1
# 95% Accuracy

class
0    500
1    268
dtype: int64

<a id="section26"></a>
## <font color="#004D7F">2.6. Correlaciones: _corr()_ </font>

Puede usar la función `corr()` para calcular una matriz de correlación. La matriz enumera todos los atributos en la parte superior y lateral, para dar correlación entre todos los pares de atributos (dos veces, porque la matriz es simétrica). Puede ver que la línea diagonal a través de la matriz desde las esquinas superior izquierda a inferior derecha de la matriz muestra una correlación perfecta de cada atributo consigo mismo.

In [17]:
# Pairwise Pearson correlations
pd.set_option('display.width', 100)
#pd.set_option('precision', 3)
correlation = data.corr(method='pearson')
print(correlation)

           preg      plas      pres      skin      test      mass      pedi       age     class
preg   1.000000  0.129459  0.141282 -0.081672 -0.073535  0.017683 -0.026205  0.544341  0.221898
plas   0.129459  1.000000  0.152590  0.057328  0.331357  0.221071  0.133163  0.263514  0.466581
pres   0.141282  0.152590  1.000000  0.207371  0.088933  0.281805  0.051436  0.239528  0.065068
skin  -0.081672  0.057328  0.207371  1.000000  0.436783  0.392573  0.154274 -0.113970  0.074752
test  -0.073535  0.331357  0.088933  0.436783  1.000000  0.197859  0.185207 -0.042163  0.130548
mass   0.017683  0.221071  0.281805  0.392573  0.197859  1.000000  0.103575  0.036242  0.292695
pedi  -0.026205  0.133163  0.051436  0.154274  0.185207  0.103575  1.000000  0.017970  0.176608
age    0.544341  0.263514  0.239528 -0.113970 -0.042163  0.036242  0.017970  1.000000  0.238356
class  0.221898  0.466581  0.065068  0.074752  0.130548  0.292695  0.176608  0.238356  1.000000


<a id="section27"></a>
## <font color="#004D7F">2.7. Asimetría: _skew()_ </font>

Puede calcular el sesgo de cada atributo utilizando la función `skew()`. El resultado de inclinación muestra una inclinación positiva (derecha) o negativa (izquierda). Los valores más cercanos a cero muestran menos sesgo.

In [18]:
# Skew for each attribute
data.skew()

preg     0.901674
plas     0.173754
pres    -1.843608
skin     0.109372
test     2.272251
mass    -0.428982
pedi     1.561978
age      1.129597
class    0.635017
dtype: float64

<div style="text-align: right"> <font size=5>
    <a href="#indice"><i class="fa fa-arrow-circle-up" aria-hidden="true" style="color:#004D7F"></i></a>
</font></div>

---

<div style="text-align: right"> <font size=6><i class="fa fa-coffee" aria-hidden="true" style="color:#004D7F"></i> </font></div>